FirebaseDatabaseUI/FirebaseIndexArray.h

```diff
@@ -0,0 +1,178 @@
+ * failed to load contents.
+ * @param array The array containing the reference.
+ * @param ref The reference that failed to load.
+ * @param index The index in the array of the reference that failed to load.
+ * @param error The error that occurred.
+ */
+- (void)array:(FirebaseIndexArray *)array
+    reference:(FIRDatabaseReference *)ref
+      atIndex:(NSUInteger)index
+didFailLoadWithError:(NSError *)error;
+
+/**
+ * Delegate method which is called whenever an object is added to a
+ * FirebaseArray. On a
+ * FirebaseArray synchronized to a Firebase reference, this corresponds to an
+ * [FIRDataEventTypeChildAdded](https://www.firebase.com/docs/ios/guide/retrieving-data.html#section-event-types)
+ * event being raised.
+ * @param ref The database reference added to the array
+ * @param index The index the reference was added at
+ */
+- (void)array:(FirebaseIndexArray *)array didAddReference:(FIRDatabaseReference *)ref atIndex:(NSUInteger)index;
+
+/**
+ * Delegate method which is called whenever an object is changed in a
+ * FirebaseArray. On a
+ * FirebaseArray synchronized to a Firebase reference, this corresponds to an
+ * [FIRDataEventTypeChildChanged](https://www.firebase.com/docs/ios/guide/retrieving-data.html#section-event-types)
+ * event being raised.
+ * @param object The database reference that changed in the array
+ * @param index The index the reference was changed at
+ */
+- (void)array:(FirebaseIndexArray *)array didChangeReference:(FIRDatabaseReference *)ref atIndex:(NSUInteger)index;
+
+/**
+ * Delegate method which is called whenever an object is removed from a
+ * FirebaseArray. On a
+ * FirebaseArray synchronized to a Firebase reference, this corresponds to an
+ * [FIRDataEventTypeChildRemoved](https://www.firebase.com/docs/ios/guide/retrieving-data.html#section-event-types)
+ * event being raised.
+ * @param object The database reference removed from the array
+ * @param index The index the reference was removed at
+ */
+- (void)array:(FirebaseIndexArray *)array didRemoveReference:(FIRDatabaseReference *)ref atIndex:(NSUInteger)index;
+
+/**
+ * Delegate method which is called whenever an object is moved within a
+ * FirebaseArray. On a
+ * FirebaseArray synchronized to a Firebase reference, this corresponds to an
+ * [FIRDataEventTypeChildMoved](https://www.firebase.com/docs/ios/guide/retrieving-data.html#section-event-types)
+ * event being raised.
+ * @param object The database reference that has moved locations
+ * @param fromIndex The index the reference is being moved from
+ * @param toIndex The index the reference is being moved to
+ */
+- (void)array:(FirebaseIndexArray *)array didMoveReference:(FIRDatabaseReference *)ref fromIndex:(NSUInteger)fromIndex toIndex:(NSUInteger)toIndex;
+
+/**
+ * Delegate method which is called whenever the backing query is canceled. This error is fatal
+ * and the index array will become unusable afterward, so please handle it appropriately.
```

Comment:
What is the appropriate handling? Guessing effectively to stop using this object